### PR TITLE
docs(examples): add widget implementation example

### PR DESCRIPTION
**Pull request by joshka**
This new example documents the various ways to implement widgets in
Ratatui. It demonstrates how to implement the `Widget` trait on a type,
a reference, and a mutable reference. It also shows how to use the
`WidgetRef` trait to render boxed widgets.

### CODE REVIEW

File: examples/widget_impl.rs
Comment (EdJoPaTo) [inline examples/widget_impl.rs:25]:
```suggestion
use self::common::Terminal;
```

File: examples/widget_impl.rs
Comment (EdJoPaTo) [inline examples/widget_impl.rs:47]: This is already a required feature for this example.

```suggestion
```

(there are multiple cases of this in this PR)

File: examples/widget_impl.rs
Comment (EdJoPaTo) [inline examples/widget_impl.rs:154]: I don't like defaults that do stuff or have side effects.

```suggestion
impl Timer {
    fn start() -> Self {
        Self {
            start: Instant::now(),
        }
    }
}
```

File: examples/widget_impl.rs
Comment (EdJoPaTo) [inline examples/widget_impl.rs:197]: Not sure that works, but it looks way less intimidating.

```suggestion
        let constraints = vec![Constraint::Length(4); self.squares.len()];
```

**Review by EdJoPaTo — COMMENTED** [review]
Quickly scrolled over it. Still against color_eyre. Also, not sure whether an example using an unstable feature is a great idea…

File: examples/widget_impl.rs
Comment (joshka) [reply examples/widget_impl.rs:47]: This is an intentional prophylactic against the regular amount "Why doesn't this compile" type questions from newer rust users.

There are multiple paths to get to seeing the examples. Newer users may not be aware of where this is set in the cargo.toml. They may be reading this on github without the context, or have copied the example code into their own projects.

File: examples/widget_impl.rs
Comment (joshka) [reply examples/widget_impl.rs:154]: Instant::now() doesn't have a side effect - it's a pure function which returns the current time.
This enables App::default(), which would otherwise need to be impl App::new().
This also isn't likely something that would trip anyone up in any way.

**Comment by joshka** [comment]
I've heard and understood your concern about color_eyre.

The examples are intentionally structured in a way that helps address common scenarios that people run into in real world building Ratatui apps. I've seen evidence (from reading a significant amount of user application code on GitHub and discord) that users often grab code from the example that most matches their particular need rather than looking at all the relevant examples. Having the necessary error handling in every example helps ensure that regardless of where they grab their code from, they're getting a reasonable basis for building their app.

This is not something I'm willing to be convinced otherwise on as the benefit to the Ratatui ecosystem is much larger than the added complexity. The only thing that would make me rethink this is if there was code that was part of Ratatui that basically did the same things. Regardless, that concern is orthogonal, and until we have something that works for that, the right approach is to harmonize the various examples with boilerplate that's easily replaced by calls to that eventual support module.

**Comment by kdheepak** [comment]
I agree about using `color_eyre` (or `anyhow`) for examples, it's something I suspect a lot of users appreciate (I know I do) and easy to remove for those that know how to do so.

**Comment by orhun** [comment]
What's the status on this? Should we get it over the finish line?

**Comment by joshka** [comment]
I believe this is ready to merge - assuming that we want to document these.
It could do with a VHS tape, but we likely need to go through and make sure all the examples are all up to date at some point, and I wouldn't block on that for this as the concept is less about the display and more about the mechanism.

**Review by orhun — APPROVED** [review]

**Pull request merged**
